gadgets/chaos_packet_drop/program.bpf.c

```diff
@@ -0,0 +1,385 @@
+} events_map SEC(".maps");
+
+GADGET_MAPITER(events_map_iter, events_map);
+
+// we use the following variables as parameters
+const volatile struct gadget_l3endpoint_t filter_ip = { 0 };
+const volatile __u16 port = 0;
+const volatile __u32 loss_percentage = 100;
+const volatile bool filter_tcp =
+	true; /* This is a boolean flag to enable filtering of TCP packets */
+const volatile bool filter_udp =
+	true; /* This is a boolean flag to enable filtering of UDP packets */
+const volatile bool ingress =
+	false; /* This is a boolean flag to enable filtering of ingress packets */
+const volatile bool egress =
+	true; /* This is a boolean flag to enable filtering of egress packets */
+
+GADGET_PARAM(filter_ip);
+GADGET_PARAM(port);
+GADGET_PARAM(loss_percentage);
+GADGET_PARAM(filter_tcp);
+GADGET_PARAM(filter_udp);
+GADGET_PARAM(ingress);
+GADGET_PARAM(egress);
+
+static __always_inline void swap_src_dst(struct event *event,
+					 struct events_map_key *key)
+{
+	struct gadget_l4endpoint_t temp = event->src;
+	event->src = key->dst;
+	key->dst = temp;
+}
+
+/* This function drops packets based on independent (Bernoulli) probability model 
+where each packet is dropped with an independent probabilty for dropping packets */
+static int rand_pkt_drop_map_update(struct event *event,
+				    struct events_map_key *key,
+				    struct sockets_key *sockets_key_for_md)
+{
+	__u32 rand_num =
+		bpf_get_prandom_u32(); // Get a random 32-bit unsigned integer
+	// Set the threshold using the loss_percentage
+	volatile __u64 threshold =
+		(volatile __u64)((volatile __u64)loss_percentage *
+				 (__u64)0xFFFFFFFF) /
+		100; // loss_percentage% of UINT32_MAX
+
+	if (ingress == true) {
+		swap_src_dst(event, key);
+	}
+	struct event *event_map_val = bpf_map_lookup_elem(
+		&events_map,
+		key); /* The events which are stored in the events_map */
+
+	if (!event)
+		return TC_ACT_OK;
+
+	if (rand_num <=
+	    (u32)threshold) // Run the code only if the random number is less than the threshold
+	{
+		if (!event_map_val) {
+			event->egress = egress;
+			event->ingress = ingress;
+			event->drop_cnt = 1;
+			/* Data collection using the socket enricher, we use the key from the map
+			to collect information regarding pid, mntns_id, tid, ppid etc */
+			sockets_key_for_md->port = key->dst.port;
+			struct sockets_value *skb_val = bpf_map_lookup_elem(
+				&gadget_sockets, sockets_key_for_md);
+			if (skb_val != NULL) {
+				event->proc.mntns_id = skb_val->mntns;
+				event->proc.pid = skb_val->pid_tgid >> 32;
+				event->proc.tid = (__u32)skb_val->pid_tgid;
+				event->proc.parent.pid = skb_val->ppid;
+				__builtin_memcpy(&event->proc.comm,
+						 skb_val->task,
+						 sizeof(event->proc.comm));
+				__builtin_memcpy(
+					&event->proc.parent.comm,
+					skb_val->ptask,
+					sizeof(event->proc.parent.comm));
+				event->proc.creds.uid = (__u32)skb_val->uid_gid;
+				event->proc.creds.gid =
+					(__u32)(skb_val->uid_gid >> 32);
+			}
```

Comment:
The `gadget_process_populate_from_socket`  could replace all of it. Please check out it's used in the trace_dns gadget.